mlflow/utils/rest_utils.py

```diff
@@ -33,22 +34,17 @@
 )
 
 
-def _get_http_response_with_retries(
-    method, url, max_retries, backoff_factor, retry_codes, **kwargs
-):
+@lru_cache(maxsize=64)
+def _get_request_session(max_retries, backoff_factor, retry_codes):
```

Comment:
We use the same max retries, backoff factor, and retry codes for nearly all requests, implying that only a single session will be used. Shall we scope this based on URL as well?

Reply:
It's very much true. And I think it is working as intended -- most of HTTP requests are sent from a single client (current runtime) to a single destination server, which should and could be considered as in the same session.

Session uses urlib3 to manage the connection pool (on TCP level) under the hood. If url differs, a new connection will be created there. I don't see a specific reason to create new sessions for that purpose.

Reply:
We make cloud storage requests to S3, Azure, GCP, etc. Is there anything problematic with reusing the same session object across all of these platforms? I'm just wondering why this wouldn't be the default behavior for requests if single session use across an entire application was recommended it most cases.

Reply:
We can certainly cache sessions based on URLs. My point is that it's less efficient, in the sense that Session is just a thin wrapper used to persist a few settings across requests. More specifically, it "Provides cookie persistence, connection-pooling, and configuration.", according to the class documentation: https://github.com/psf/requests/blob/main/requests/sessions.py#L327

From https://github.com/psf/requests/blob/main/requests/sessions.py#L343, among all stuff a Session persists currently
```
        'headers', 'cookies', 'auth', 'proxies', 'hooks', 'params', 'verify',
        'cert', 'adapters', 'stream', 'trust_env', 'max_redirects',
```
the only setting that's being persisted in our use case is the `adapters` (which could change depending on the retry logic), which does the TCP connection pooling and re-use.

In the future, if we decide to persist auth and cookies, it probably makes sense to change the cache logic to take these into account, which should happen automatically since auth and cookie settings have to be passed into the session generator function `_get_request_session` as part of the argument, which will change the session cache key signature.

Reply:
Understood! Thanks for the clarification. Can we manually confirm that logging both models and metadata (metrics, params, etc) works properly on Databricks before merging? I see we've already tested metrics.

Reply:
Tested on Databricks with logging metrics, params, artifacts and models repeatedly. No problem was observed. We see some performance improvements, especially on repeated small requests such as metric logging.